llama.h

```diff
@@ -853,7 +856,7 @@ extern "C" {
 
     // Special tokens
     LLAMA_API llama_token llama_token_bos(const struct llama_model * model); // beginning-of-sentence
-    LLAMA_API llama_token llama_token_eos(const struct llama_model * model); // end-of-sentence
+    LLAMA_API void llama_token_eos(const struct llama_model * model, llama_token* token_list); // end-of-sentence
```

Comment:
This does not look great. Could you summarize why these models require multiple EOS tokens?

If it is really needed, we might introduce:

```c
LLAMA_API llama_token llama_token_eos_ith(const struct llama_model * model, int i);
```

But first need to understand if this is really necessary

Reply:
@ggerganov Normally, a chat looks like this:

```
<|user|>Who are you<|end|><|assistant|>I'm am AI<|end|>
```

But because this model does not have EOT (end-of-turn) token, the chat looks like this:

```
<|user|>Who are you<|assistant|>I'm am AI<|user|>
```

So the EOT should be `<|user|>`. The list of EOT token contains `<|user|>`, `<|observation|>` and `<|endoftext|>`

But after second thought @youth123 , I realize that can we set `<|endoftext|>` as EOS and `<|user|>` as EOG? We can ignore `<|observation|>` for now, because it's inside generated text (I think?). So finally no need for adding list of EOS (sorry for not thinking about this earlier)

Reply:
Multiple EOS tokens are used to enable the language model to support function calls and tool calls, making it easier to distinguish between roles. I will implement it according to the solution provided by @ngxson . However, for future scalability, I still hope that the official support for multiple EOS tokens will be provided.

Reply:
@youth123 Yeah I'd agree that multiple EOT is very useful for tool calls. However, `llama-cli` and `llama-server` don't yet have support for tool calls. I gave a try with "functionary" a while ago, but I'm not very satisfied with the implementation so it's not get merged. Hopefully we will be able to revisit it in the future.

Reply:
The `llama_token_is_eog()` was introduced to solve this exact problem:

https://github.com/ggerganov/llama.cpp/blob/168b6f5a4cc52eca11c6bd20b2867e2f603134ef/include/llama.h#L847-L850

You need to list all such tokens here:

https://github.com/ggerganov/llama.cpp/blob/168b6f5a4cc52eca11c6bd20b2867e2f603134ef/src/llama.cpp#L2339-L2347

For example, add `id special_obs_id = -1;`

In `llm_load_vocab()` initialize from the KV header:

https://github.com/ggerganov/llama.cpp/blob/168b6f5a4cc52eca11c6bd20b2867e2f603134ef/src/llama.cpp#L5063-L5078

Or try to match them by name (not recommended):

https://github.com/ggerganov/llama.cpp/blob/168b6f5a4cc52eca11c6bd20b2867e2f603134ef/src/llama.cpp#L5107-L5128

Extend `llama_token_is_eog()`:

```c++
bool llama_token_is_eog(const struct llama_model * model, llama_token token) {
    return token != -1 && (
        token == llama_token_eos(model) ||
        token == llama_token_eot(model) ||
        token == llama_token_obs(model)
    );
}
```

Reply:
Okay, thank you. I will follow these instructions to make the changes.

Reply:
I have already set eot and eos in convert-hf-gguf. 
https://github.com/youth123/llama.cpp/blob/0595f03dd171cdad54aafe79199f223d7a2e3f0b/convert-hf-to-gguf.py#L3052C9-L3053C89
``` python
        special_vocab._set_special_token("eos", tokenizer.get_added_vocab()["<|endoftext|>"])
        special_vocab._set_special_token("eot", tokenizer.get_added_vocab()["<|user|>"])
```
Then, when I merged the latest code, I found a bug. There is a risk of type conversion overflow in the llama_chat_apply_template function in common.cpp. When tmpl does not match the current template, res will become negative and then be forcibly converted to uint.
``` c++
std::string llama_chat_apply_template(const struct llama_model * model,
        const std::string & tmpl,
        const std::vector<llama_chat_msg> & msgs,
        bool add_ass) {
    int alloc_size = 0;
    std::vector<llama_chat_message> chat;
    for (auto & msg : msgs) {
        chat.push_back({msg.role.c_str(), msg.content.c_str()});
        alloc_size += (msg.role.size() + msg.content.size()) * 1.25;
    }

    const char * ptr_tmpl = tmpl.empty() ? nullptr : tmpl.c_str();
    std::vector<char> buf(alloc_size);

    // run the first time to get the total output length
    int32_t res = llama_chat_apply_template(model, ptr_tmpl, chat.data(), chat.size(), add_ass, buf.data(), buf.size());

    // if it turns out that our buffer is too small, we resize it
    if ((size_t) res > buf.size()) {
        buf.resize(res);
        res = llama_chat_apply_template(model, ptr_tmpl, chat.data(), chat.size(), add_ass, buf.data(), buf.size());
    }

    std::string formatted_chat(buf.data(), res);
    return formatted_chat;
}
```

Reply:
> Then, when I merged the latest code, I found a bug. There is a risk of type conversion overflow in the llama_chat_apply_template function in common.cpp. When tmpl does not match the current template, res will become negative and then be forcibly converted to uint.

`std::string llama_chat_apply_template()` is a thin wrapper for the underlaying c style `llama_chat_apply_template`, so it's required to use `llama_chat_verify_template` to check if the template is supported or not (if not, we exit). You can see this check in `gpt_params_find_arg`

But I think it's OK to add a bit redundant check in the cpp function here, can be as simple as

```cpp
if (res < 0) {
  throw std::runtime_error("template is not supported");
}
```

Reply:
Ok, I understand. However, it seems that the input parameters must include chat-template; otherwise, this check will not be effective.